File: Tests/MoongateCoreTests/SubtitleSourceDecisionEngineTests.swift
import XCTest
@testable import MoongateCore

final class SubtitleSourceDecisionEngineTests: XCTestCase {

    // MARK: - Helpers

    private func assessment(
        id: String = "c",
        kind: SubtitleSourceKind,
        score: Double,
        gateUsable: Bool,
        verdict: SubtitleQualityVerdict,
        hasFile: Bool = true
    ) -> SubtitleSourceDecisionEngine.Assessment {
        SubtitleSourceDecisionEngine.Assessment(
            candidateID: id,
            kind: kind,
            languageCode: "ja",
            score: score,
            verdict: verdict,
            gateUsable: gateUsable,
            gateReasons: gateUsable ? [] : [.tooFewCues],
            reasons: [],
            report: nil,
            hasFile: hasFile
        )
    }

    // MARK: - generationPlan: gate-only, no conflation

    func testAutoBestKeepsHealthyPlatformWithoutGenerating() {
        let platform = assessment(kind: .platformAuto, score: 75, gateUsable: true, verdict: .good)
        let plan = SubtitleSourceDecisionEngine.generationPlan(
            policy: .autoBest, platform: platform, localASRAvailable: true, cloudASRAvailable: true)
        XCTAssertEqual(plan, .none)
    }

    func testAutoBestRegeneratesWhenGateUnusableAndLocalAvailable() {
        let platform = assessment(kind: .platformAuto, score: 30, gateUsable: false, verdict: .unusable)
        let plan = SubtitleSourceDecisionEngine.generationPlan(
            policy: .autoBest, platform: platform, localASRAvailable: true, cloudASRAvailable: false)
        XCTAssertEqual(plan, .generateLocalASRThenChoose)
    }

    func testAutoBestKeepsPlatformWithReasonsWhenGateUnusableButNoGenerator() {
        let platform = assessment(kind: .platformAuto, score: 30, gateUsable: false, verdict: .unusable)
        let plan = SubtitleSourceDecisionEngine.generationPlan(
            policy: .autoBest, platform: platform, localASRAvailable: false, cloudASRAvailable: false)
        XCTAssertEqual(plan, .keepPlatformRecordReasons([.tooFewCues]))
    }

    /// 质量地板：门 usable 但 verdict 低于 usable(lowConfidence) 时仍重生成——行为与旧
    /// `|| score <= .lowConfidence` 一致，但现在是单类型、具名、可测，不再混 OR。
    func testAutoBestRegeneratesWhenGateUsableButVerdictBelowFloor() {
        let platform = assessment(kind: .platformAuto, score: 45, gateUsable: true, verdict: .lowConfidence)
        let plan = SubtitleSourceDecisionEngine.generationPlan(
            policy: .autoBest, platform: platform, localASRAvailable: true, cloudASRAvailable: false)
        XCTAssertEqual(plan, .generateLocalASRThenChoose)
    }

    /// 反例：门 usable 且 verdict 达 usable → 不生成（不会因 sub-fatal 噪声白跑 Whisper）。
    func testAutoBestDoesNotRegenerateWhenUsableVerdict() {
        let platform = assessment(kind: .platformAuto, score: 60, gateUsable: true, verdict: .usable)
        let plan = SubtitleSourceDecisionEngine.generationPlan(
            policy: .autoBest, platform: platform, localASRAvailable: true, cloudASRAvailable: false)
        XCTAssertEqual(plan, .none)
    }

    func testPlatformPoliciesNeverGenerate() {
        let platform = assessment(kind: .platformAuto, score: 30, gateUsable: false, verdict: .unusable)
        for policy in [SubtitleSourcePolicy.forcePlatform, .preferPlatform, .cloudASR, .importedFile] {
            let plan = SubtitleSourceDecisionEngine.generationPlan(
                policy: policy, platform: platform, localASRAvailable: true, cloudASRAvailable: true)
            XCTAssertEqual(plan, .none, "\(policy) should never generate local ASR")
        }
    }

    func testForceLocalASRAlwaysGenerates() {
        let platform = assessment(kind: .platformAuto, score: 90, gateUsable: true, verdict: .excellent)
        let plan = SubtitleSourceDecisionEngine.generationPlan(
            policy: .forceLocalASR, platform: platform, localASRAvailable: true, cloudASRAvailable: false)
        XCTAssertEqual(plan, .generateLocalASRThenChoose)
    }

    func testPreferLocalASRGeneratesOnlyWhenPlatformUnusable() {
        let healthy = assessment(kind: .platformAuto, score: 80, gateUsable: true, verdict: .good)
        XCTAssertEqual(
            SubtitleSourceDecisionEngine.generationPlan(
                policy: .preferLocalASR, platform: healthy, localASRAvailable: true, cloudASRAvailable: false),
            .none)
        let bad = assessment(kind: .platformAuto, score: 30, gateUsable: false, verdict: .unusable)
        XCTAssertEqual(
            SubtitleSourceDecisionEngine.generationPlan(
                policy: .preferLocalASR, platform: bad, localASRAvailable: true, cloudASRAvailable: false),
            .generateLocalASRThenChoose)
    }

    // MARK: - choose: tie-break prefers more-trusted source

    func testChooseTieBreakPrefersLowerSourceKindRank() {
        let manual = assessment(id: "m", kind: .manual, score: 70, gateUsable: true, verdict: .good)
        let local = assessment(id: "l", kind: .localASR, score: 70, gateUsable: true, verdict: .good)
        let winner = SubtitleSourceDecisionEngine.choose(
            policy: .autoBest, assessments: [manual, local], selectableIDs: ["m", "l"])
        XCTAssertEqual(winner, "m")
        // 顺序无关
        let winner2 = SubtitleSourceDecisionEngine.choose(
            policy: .autoBest, assessments: [local, manual], selectableIDs: ["m", "l"])
        XCTAssertEqual(winner2, "m")
    }

    func testChooseAutoBestPicksHigherScore() {
        let manual = assessment(id: "m", kind: .manual, score: 60, gateUsable: true, verdict: .usable)
        let local = assessment(id: "l", kind: .localASR, score: 80, gateUsable: true, verdict: .good)
        let winner = SubtitleSourceDecisionEngine.choose(
            policy: .autoBest, assessments: [manual, local], selectableIDs: ["m", "l"])
        XCTAssertEqual(winner, "l")
    }

    func testChooseForcePlatformPicksPlatformEvenWhenLocalScoresHigher() {
        let platform = assessment(id: "p", kind: .platformAuto, score: 55, gateUsable: true, verdict: .usable)
        let local = assessment(id: "l", kind: .localASR, score: 95, gateUsable: true, verdict: .excellent)
        let winner = SubtitleSourceDecisionEngine.choose(
            policy: .forcePlatform, assessments: [platform, local], selectableIDs: ["p", "l"])
        XCTAssertEqual(winner, "p")
    }

    func testChooseSkipsNonSelectableCandidates() {
        let pending = assessment(id: "pending", kind: .localASR, score: 0, gateUsable: false, verdict: .unusable, hasFile: false)
        let platform = assessment(id: "p", kind: .platformAuto, score: 60, gateUsable: true, verdict: .usable)
        let winner = SubtitleSourceDecisionEngine.choose(
            policy: .autoBest, assessments: [pending, platform], selectableIDs: ["p"])
        XCTAssertEqual(winner, "p")
    }

    // MARK: - assess: one gate run, authoritative gateUsable

    func testAssessHealthyJapaneseIsGateUsable() throws {
        let url = try writeSRT(name: "local-asr.ja.srt", texts: [
            "今日は楽しいお祭りの日です", "みんなでチョコバナナを食べよう", "ソースせんべいも買ってきたよ",
            "お風呂はとても気持ちいいね", "ありがとうと言われるとうれしい", "風が涼しくて気持ちいい",
            "友だちと一緒に歩いている", "次はくじ引きをやってみよう", "小さな声でもちゃんと聞こえる", "また明日も遊びに来よう",
        ])
        let assessment = SubtitleSourceDecisionEngine.assess(
            candidate: SubtitleSourceCandidate(id: "l", kind: .localASR, languageCode: "ja", displayName: "L",
                                               fileURL: url, isGenerated: true, provider: "whisper.cpp"),
            requestedSourceLanguageCode: "ja", videoDurationSeconds: nil)
        XCTAssertTrue(assessment.gateUsable)
        XCTAssertTrue(assessment.gateReasons.isEmpty)
        XCTAssertGreaterThanOrEqual(assessment.verdict, .usable)
    }

    func testAssessTooFewCuesIsGateUnusable() throws {
        let url = try writeSRT(name: "auto.ja.srt", texts: ["短い", "字幕", "です"])
        let assessment = SubtitleSourceDecisionEngine.assess(
            candidate: SubtitleSourceCandidate(id: "p", kind: .platformAuto, languageCode: "ja", displayName: "P",
                                               fileURL: url, isGenerated: true, provider: "yt-dlp"),
            requestedSourceLanguageCode: "ja", videoDurationSeconds: nil)
        XCTAssertFalse(assessment.gateUsable)
        XCTAssertTrue(assessment.gateReasons.contains(.tooFewCues))
        // 不变式：gate 不可用 ⇒ verdict <= lowConfidence
        XCTAssertLessThanOrEqual(assessment.verdict, .lowConfidence)
    }

    func testAssessMissingFileIsGateUnusable() {
        let assessment = SubtitleSourceDecisionEngine.assess(
            candidate: SubtitleSourceCandidate(id: "pending", kind: .localASR, languageCode: "ja", displayName: "L",
                                               fileURL: nil, isGenerated: false, provider: "whisper.cpp"),
            requestedSourceLanguageCode: "ja", videoDurationSeconds: nil)
        XCTAssertFalse(assessment.gateUsable)
        XCTAssertFalse(assessment.hasFile)
    }

    func testEngineConstantsMatchCrossPlatformFixture() throws {
        let fixture = try loadFixtureSection("subtitleSourceDecision")
        func section(_ key: String) throws -> [String: Any] {
            try XCTUnwrap(fixture[key] as? [String: Any], "fixture missing \(key)")
        }
        // baseScore（来源先验，真值在 SubtitleQualityScorer）
        let base = try section("baseScore")
        let baseExpect: [(SubtitleSourceKind, String)] = [
            (.manual, "manual"), (.importedFile, "importedFile"), (.hlsManifest, "hlsManifest"),
            (.cloudASR, "cloudASR"), (.platformAuto, "platformAuto"), (.localASR, "localASR"),
        ]
        for (kind, key) in baseExpect {
            let want = try XCTUnwrap((base[key] as? NSNumber)?.doubleValue, "baseScore.\(key)")
            XCTAssertEqual(SubtitleQualityScorer.baseScore(for: kind), want, "baseScore.\(key)")
        }
        // sourceKindRank（tie-break 顺序，真值在引擎）
        let rank = try section("sourceKindRank")
        for (kind, key) in baseExpect {
            let want = try XCTUnwrap((rank[key] as? NSNumber)?.intValue, "sourceKindRank.\(key)")
            XCTAssertEqual(SubtitleSourceDecisionEngine.sourceKindRank(kind), want, "sourceKindRank.\(key)")
        }
        // policyBoost：prefer / force
        let boost = try section("policyBoost")
        let preferWant = try XCTUnwrap((boost["prefer"] as? NSNumber)?.doubleValue)
        let forceWant = try XCTUnwrap((boost["force"] as? NSNumber)?.doubleValue)
        XCTAssertEqual(SubtitleSourceDecisionEngine.policyBoost(.platformAuto, .preferPlatform), preferWant)
        XCTAssertEqual(SubtitleSourceDecisionEngine.policyBoost(.localASR, .preferLocalASR), preferWant)
        XCTAssertEqual(SubtitleSourceDecisionEngine.policyBoost(.localASR, .forceLocalASR), forceWant)
        // verdictThresholds
        let vt = try section("verdictThresholds")
        XCTAssertEqual(SubtitleQualityScorer.verdictThresholds.excellent, try XCTUnwrap((vt["excellent"] as? NSNumber)?.doubleValue))
        XCTAssertEqual(SubtitleQualityScorer.verdictThresholds.good, try XCTUnwrap((vt["good"] as? NSNumber)?.doubleValue))
        XCTAssertEqual(SubtitleQualityScorer.verdictThresholds.usable, try XCTUnwrap((vt["usable"] as? NSNumber)?.doubleValue))
        XCTAssertEqual(SubtitleQualityScorer.verdictThresholds.lowConfidence, try XCTUnwrap((vt["lowConfidence"] as? NSNumber)?.doubleValue))
        // autoBestRegenerateBelow
        XCTAssertEqual(fixture["autoBestRegenerateBelow"] as? String, SubtitleSourceDecisionEngine.autoBestRegenerateBelow.rawValue)
    }

    private func loadFixtureSection(_ section: String) throws -> [String: Any] {
        let url = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .appendingPathComponent("Tests/fixtures/whisper-timing-constants.json")
        let data = try Data(contentsOf: url)
        let fixture = try XCTUnwrap(try JSONSerialization.jsonObject(with: data) as? [String: Any])
        return try XCTUnwrap(fixture[section] as? [String: Any], "fixture missing section \(section)")
    }

    // MARK: - SRT helper

    private func writeSRT(name: String, texts: [String]) throws -> URL {
        let dir = FileManager.default.temporaryDirectory
            .appendingPathComponent("moongate-engine-tests-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let url = dir.appendingPathComponent(name)
        let body = texts.enumerated().map { index, text in
            let start = index * 2_000
            let end = start + 1_500
            return "\(index + 1)\n\(timestamp(start)) --> \(timestamp(end))\n\(text)"
        }.joined(separator: "\n\n")
        try body.write(to: url, atomically: true, encoding: .utf8)
        return url
    }

    private func timestamp(_ totalMs: Int) -> String {
        String(format: "%02d:%02d:%02d,%03d",
               totalMs / 3_600_000, (totalMs % 3_600_000) / 60_000, (totalMs % 60_000) / 1000, totalMs % 1000)
    }
}

File: windows/MoongateCore.Tests/SubtitleSourceDecisionEngineTests.cs
using Moongate.Core;

namespace MoongateCore.Tests;

public sealed class SubtitleSourceDecisionEngineTests : IDisposable
{
    private readonly string _directory = Path.Combine(Path.GetTempPath(), "mg-engine-" + Guid.NewGuid().ToString("N"));

    public SubtitleSourceDecisionEngineTests() => Directory.CreateDirectory(_directory);

    public void Dispose()
    {
        try { Directory.Delete(_directory, recursive: true); } catch { /* ignore */ }
    }

    private static SubtitleSourceDecisionEngine.Assessment Assessment(
        SubtitleSourceKind kind,
        double score,
        bool gateUsable,
        SubtitleQualityVerdict verdict,
        string id = "c",
        bool hasFile = true) =>
        new(id, kind, "ja", score, verdict, gateUsable,
            gateUsable ? [] : [PlatformSubtitleQualityGate.Reason.TooFewCues],
            [], null, hasFile);

    // ---- generationPlan: gate-only, no conflation ----

    [Fact]
    public void AutoBestKeepsHealthyPlatformWithoutGenerating()
    {
        var platform = Assessment(SubtitleSourceKind.PlatformAuto, 75, true, SubtitleQualityVerdict.Good);
        var plan = SubtitleSourceDecisionEngine.GenerationPlan(SubtitleSourcePolicy.AutoBest, platform, true, true);
        Assert.Equal(SubtitleSourceDecisionEngine.AsrPlanKind.None, plan.Kind);
    }

    [Fact]
    public void AutoBestRegeneratesWhenGateUnusableAndLocalAvailable()
    {
        var platform = Assessment(SubtitleSourceKind.PlatformAuto, 30, false, SubtitleQualityVerdict.Unusable);
        var plan = SubtitleSourceDecisionEngine.GenerationPlan(SubtitleSourcePolicy.AutoBest, platform, true, false);
        Assert.Equal(SubtitleSourceDecisionEngine.AsrPlanKind.GenerateLocalAsrThenChoose, plan.Kind);
    }

    [Fact]
    public void AutoBestKeepsPlatformWithReasonsWhenNoGenerator()
    {
        var platform = Assessment(SubtitleSourceKind.PlatformAuto, 30, false, SubtitleQualityVerdict.Unusable);
        var plan = SubtitleSourceDecisionEngine.GenerationPlan(SubtitleSourcePolicy.AutoBest, platform, false, false);
        Assert.Equal(SubtitleSourceDecisionEngine.AsrPlanKind.KeepPlatformRecordReasons, plan.Kind);
        Assert.Contains(PlatformSubtitleQualityGate.Reason.TooFewCues, plan.Reasons);
    }

    [Fact]
    public void AutoBestRegeneratesWhenGateUsableButVerdictBelowFloor()
    {
        var platform = Assessment(SubtitleSourceKind.PlatformAuto, 45, true, SubtitleQualityVerdict.LowConfidence);
        var plan = SubtitleSourceDecisionEngine.GenerationPlan(SubtitleSourcePolicy.AutoBest, platform, true, false);
        Assert.Equal(SubtitleSourceDecisionEngine.AsrPlanKind.GenerateLocalAsrThenChoose, plan.Kind);
    }

    [Fact]
    public void AutoBestDoesNotRegenerateWhenUsableVerdict()
    {
        var platform = Assessment(SubtitleSourceKind.PlatformAuto, 60, true, SubtitleQualityVerdict.Usable);
        var plan = SubtitleSourceDecisionEngine.GenerationPlan(SubtitleSourcePolicy.AutoBest, platform, true, false);
        Assert.Equal(SubtitleSourceDecisionEngine.AsrPlanKind.None, plan.Kind);
    }

    [Theory]
    [InlineData(SubtitleSourcePolicy.ForcePlatform)]
    [InlineData(SubtitleSourcePolicy.PreferPlatform)]
    [InlineData(SubtitleSourcePolicy.CloudAsr)]
    [InlineData(SubtitleSourcePolicy.ImportedFile)]
    public void PlatformPoliciesNeverGenerate(SubtitleSourcePolicy policy)
    {
        var platform = Assessment(SubtitleSourceKind.PlatformAuto, 30, false, SubtitleQualityVerdict.Unusable);
        var plan = SubtitleSourceDecisionEngine.GenerationPlan(policy, platform, true, true);
        Assert.Equal(SubtitleSourceDecisionEngine.AsrPlanKind.None, plan.Kind);
    }

    [Fact]
    public void ForceLocalAsrAlwaysGenerates()
    {
        var platform = Assessment(SubtitleSourceKind.PlatformAuto, 90, true, SubtitleQualityVerdict.Excellent);
        var plan = SubtitleSourceDecisionEngine.GenerationPlan(SubtitleSourcePolicy.ForceLocalAsr, platform, true, false);
        Assert.Equal(SubtitleSourceDecisionEngine.AsrPlanKind.GenerateLocalAsrThenChoose, plan.Kind);
    }

    [Fact]
    public void PreferLocalAsrGeneratesOnlyWhenPlatformUnusable()
    {
        var healthy = Assessment(SubtitleSourceKind.PlatformAuto, 80, true, SubtitleQualityVerdict.Good);
        Assert.Equal(SubtitleSourceDecisionEngine.AsrPlanKind.None,
            SubtitleSourceDecisionEngine.GenerationPlan(SubtitleSourcePolicy.PreferLocalAsr, healthy, true, false).Kind);
        var bad = Assessment(SubtitleSourceKind.PlatformAuto, 30, false, SubtitleQualityVerdict.Unusable);
        Assert.Equal(SubtitleSourceDecisionEngine.AsrPlanKind.GenerateLocalAsrThenChoose,
            SubtitleSourceDecisionEngine.GenerationPlan(SubtitleSourcePolicy.PreferLocalAsr, bad, true, false).Kind);
    }

    // ---- choose: tie-break prefers more-trusted source ----

    [Fact]
    public void ChooseTieBreakPrefersLowerSourceKindRank()
    {
        var manual = Assessment(SubtitleSourceKind.Manual, 70, true, SubtitleQualityVerdict.Good, "m");
        var local = Assessment(SubtitleSourceKind.LocalAsr, 70, true, SubtitleQualityVerdict.Good, "l");
        Assert.Equal("m", SubtitleSourceDecisionEngine.Choose(
            SubtitleSourcePolicy.AutoBest, [manual, local], new HashSet<string> { "m", "l" }));
        Assert.Equal("m", SubtitleSourceDecisionEngine.Choose(
            SubtitleSourcePolicy.AutoBest, [local, manual], new HashSet<string> { "m", "l" }));
    }

    [Fact]
    public void ChooseAutoBestPicksHigherScore()
    {
        var manual = Assessment(SubtitleSourceKind.Manual, 60, true, SubtitleQualityVerdict.Usable, "m");
        var local = Assessment(SubtitleSourceKind.LocalAsr, 80, true, SubtitleQualityVerdict.Good, "l");
        Assert.Equal("l", SubtitleSourceDecisionEngine.Choose(
            SubtitleSourcePolicy.AutoBest, [manual, local], new HashSet<string> { "m", "l" }));
    }

    [Fact]
    public void ChooseForcePlatformPicksPlatformEvenWhenLocalScoresHigher()
    {
        var platform = Assessment(SubtitleSourceKind.PlatformAuto, 55, true, SubtitleQualityVerdict.Usable, "p");
        var local = Assessment(SubtitleSourceKind.LocalAsr, 95, true, SubtitleQualityVerdict.Excellent, "l");
        Assert.Equal("p", SubtitleSourceDecisionEngine.Choose(
            SubtitleSourcePolicy.ForcePlatform, [platform, local], new HashSet<string> { "p", "l" }));
    }

    [Fact]
    public void ChooseSkipsNonSelectableCandidates()
    {
        var pending = Assessment(SubtitleSourceKind.LocalAsr, 0, false, SubtitleQualityVerdict.Unusable, "pending", hasFile: false);
        var platform = Assessment(SubtitleSourceKind.PlatformAuto, 60, true, SubtitleQualityVerdict.Usable, "p");
        Assert.Equal("p", SubtitleSourceDecisionEngine.Choose(
            SubtitleSourcePolicy.AutoBest, [pending, platform], new HashSet<string> { "p" }));
    }

    // ---- assess: one gate run, authoritative gateUsable ----

    [Fact]
    public void AssessHealthyJapaneseIsGateUsable()
    {
        var path = WriteSrt("local-asr.ja.srt", new[]
        {
            "今日は楽しいお祭りの日です", "みんなでチョコバナナを食べよう", "ソースせんべいも買ってきたよ",
            "お風呂はとても気持ちいいね", "ありがとうと言われるとうれしい", "風が涼しくて気持ちいい",
            "友だちと一緒に歩いている", "次はくじ引きをやってみよう", "小さな声でもちゃんと聞こえる", "また明日も遊びに来よう",
        });
        var assessment = SubtitleSourceDecisionEngine.Assess(
            new SubtitleSourceCandidate("l", SubtitleSourceKind.LocalAsr, "ja", "L", path, true, "whisper.cpp"),
            "ja", null);
        Assert.True(assessment.GateUsable);
        Assert.Empty(assessment.GateReasons);
        Assert.True(assessment.Verdict >= SubtitleQualityVerdict.Usable);
    }

    [Fact]
    public void AssessTooFewCuesIsGateUnusable()
    {
        var path = WriteSrt("auto.ja.srt", new[] { "短い", "字幕", "です" });
        var assessment = SubtitleSourceDecisionEngine.Assess(
            new SubtitleSourceCandidate("p", SubtitleSourceKind.PlatformAuto, "ja", "P", path, true, "yt-dlp"),
            "ja", null);
        Assert.False(assessment.GateUsable);
        Assert.Contains(PlatformSubtitleQualityGate.Reason.TooFewCues, assessment.GateReasons);
        Assert.True(assessment.Verdict <= SubtitleQualityVerdict.LowConfidence);
    }

    [Fact]
    public void AssessMissingFileIsGateUnusable()
    {
        var assessment = SubtitleSourceDecisionEngine.Assess(
            new SubtitleSourceCandidate("pending", SubtitleSourceKind.LocalAsr, "ja", "L", null, false, "whisper.cpp"),
            "ja", null);
        Assert.False(assessment.GateUsable);
        Assert.False(assessment.HasFile);
    }

    private string WriteSrt(string name, IReadOnlyList<string> texts)
    {
        var cues = texts.Select((text, index) =>
            new SubtitleCue(index + 1,
                SrtTools.SecondsToSrtTime(index * 2.0),
                SrtTools.SecondsToSrtTime(index * 2.0 + 1.5),
                text, [])).ToArray();
        var path = Path.Combine(_directory, name);
        File.WriteAllText(path, SrtTools.SerializeSrt(cues));
        return path;
    }
}
